qiskit/ml/circuit/library/raw_feature_vector.py

```diff
@@ -0,0 +1,160 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The raw feature vector circuit."""
+
+from typing import Set, List, Optional
+import numpy as np
+from qiskit.circuit import QuantumRegister, ParameterVector, ParameterExpression, Gate
+from qiskit.circuit.library import BlueprintCircuit
+
+
+class RawFeatureVector(BlueprintCircuit):
+    """The raw feature vector circuit."""
+
+    def __init__(self, feature_dimension: Optional[int]) -> None:
+        """
+        Args:
+            feature_dimension: The feature dimension and number of qubits.
+
+        """
+        super().__init__()
+
+        self._num_qubits = None
+        self._parameters = None
+
+        if feature_dimension:
+            self.feature_dimension = feature_dimension
+
+    def _build(self):
+        super()._build()
+
+        # if the parameters are fully specified, use the initialize instruction
+        if len(self.parameters) == 0:
+            self.initialize(self._parameters, self.qubits)  # pylint: disable=no-member
+
+        # otherwise get a gate that acts as placeholder
+        else:
+            placeholder = Gate('Raw', self.num_qubits, self._parameters[:], label='Raw')
+            self.append(placeholder, self.qubits)
+
+    def _check_configuration(self, raise_on_failure=True):
+        pass
+
+    @property
+    def num_qubits(self) -> int:
+        """Returns the number of qubits in this circuit.
+
+        Returns:
+            The number of qubits.
+        """
+        return self._num_qubits if self._num_qubits is not None else 0
+
+    @num_qubits.setter
+    def num_qubits(self, num_qubits: int) -> None:
+        """Set the number of qubits for the n-local circuit.
+
+        Args:
+            The new number of qubits.
+        """
+        if self._num_qubits != num_qubits:
+            # invalidate the circuit
+            self._invalidate()
+            self._num_qubits = num_qubits
+            self._parameters = list(ParameterVector('p', length=self.feature_dimension))
+            self.add_register(QuantumRegister(self.num_qubits, 'q'))
+
+    @property
+    def feature_dimension(self) -> int:
+        """Return the feature dimension.
+
+        Returns:
+            The feature dimension, which is ``2 ** num_qubits``.
+        """
+        return 2 ** self.num_qubits
+
+    @feature_dimension.setter
+    def feature_dimension(self, feature_dimension: int) -> None:
+        """Set the feature dimension.
+
+        Args:
+            feature_dimension: The new feature dimension. Must be a power of 2.
+
+        Raises:
+            ValueError: If ``feature_dimension`` is not a power of 2.
+        """
+        num_qubits = np.log2(feature_dimension)
+        if self._num_qubits is None or num_qubits != self._num_qubits:
+            if int(num_qubits) != num_qubits:
+                raise ValueError('feature_dimension must be a power of 2!')
```

Comment:
This error may end up being confusing. It seems feature_dimension and/or num_qubits can be set, A) we need feature dimension to be a power of 2, but B) it must match whatever the num_qubits was set to and if that aspect fails it still complains about power of 2. 

Also should the num qubits setter do a sanity check again feature dimension which it does not seem to do above.

Reply:
Which part is confusing? 🙂 This should just check whether the feature dimension is a power of 2.

It doesn't need to match the number of qubits, since it's a blueprintcircuit the number of qubits is mutable. So I could do
```
qc = RawFeatureVector(4)  # qc with 2 qubits
qc.feature_dimension = 8  # qc has now 3 qubits
```

Reply:
Ok, I guess because it was already inside the if  having done this ` or num_qubits != self._num_qubits:` that confused me around qubit mismatching. Can I suggest doing it this way instead without the nested if. This way the check for validity is separate from the check if the circuit needs invalidation. Also the test against self._qubits being None is not needed I think as the test for the new value (which should not be None) would still mismatch (the only minor thing is maybe you don't need to invalidate if its still None but thats minor I guess)
```
        num_qubits = np.log2(feature_dimension)
        if int(num_qubits) != num_qubits:
              raise ValueError('feature_dimension must be a power of 2!')

        if (int)num_qubits != self._num_qubits:
            self._invalidate()
            self.num_qubits = int(num_qubits)
``

Reply:
Done 👍